### ISSUE TITLE
Notifications: Inconsistent State

### ISSUE THREAD

**Issue by jleandroperez** [wordpress-mobile/WordPress-iOS#6341]
### Details:
1. Launch WPiOS and open the Notifications tab
2. Send the app to Background
3. Open the System Settings, and bump up the date +1 day
4. With the app in BG, receive a new notification

I managed to trigger this just once (can't repro reliably, there has to be another factor). Console logs pasted below:

After this error was displayed in the console, the FRC ended up out of sync with the Core Data state, and pressing any random UITableViewCell would push the incorrect results.

Thanks a whole lot to @meremagee for helping me spot this one!!!

```
2016-12-16 18:04:35:515 WordPress[2579:786921] Current Application state: 2
2016-12-16 18:04:35.520445 WordPress[2579:789692] Current Application state: 2
2016-12-16 18:04:35:530 WordPress[2579:786921] Running Notifications Background Fetch...
2016-12-16 18:04:35.531621 WordPress[2579:788881] Running Notifications Background Fetch...
2016-12-16 18:04:36.685568 WordPress[2579:786921] *** Assertion failure in -[UITableView _endCellAnimationsWithContext:], /BuildRoot/Library/Caches/com.apple.xbs/Sources/UIKit/UIKit-3600.6.22/UITableView.m:1396
2016-12-16 18:04:36.689725 WordPress[2579:786921] [error] error: Serious application error.  An exception was caught from the delegate of NSFetchedResultsController during a call to -controllerDidChangeContent:.  attempt to insert row 3 into section 0, but there are only 3 rows in section 0 after the update with userInfo (null)
CoreData: error: Serious application error.  An exception was caught from the delegate of NSFetchedResultsController during a call to -controllerDidChangeContent:.  attempt to insert row 3 into section 0, but there are only 3 rows in section 0 after the update with userInfo (null)
2016-12-16 18:04:36.697382 WordPress[2579:789689] Finished Notifications Background Fetch!
2016-12-16 18:04:36:696 WordPress[2579:786921] Finished Notifications Background Fetch!
```

**Comment by jleandroperez** [comment]
Aaaand broken yet again. Same steps as above, i just left the app sitting for a while, and got a bunch of notifications in the meantime.

```
2016-12-18 18:23:49.596194 WordPress[2598:792178] *** Assertion failure in -[UITableView _endCellAnimationsWithContext:], /BuildRoot/Library/Caches/com.apple.xbs/Sources/UIKit/UIKit-3600.6.22/UITableView.m:1610
2016-12-18 18:23:49.598760 WordPress[2598:792178] [error] error: Serious application error.  An exception was caught from the delegate of NSFetchedResultsController during a call to -controllerDidChangeContent:.  Invalid update: invalid number of rows in section 0.  The number of rows contained in an existing section after the update (83) must be equal to the number of rows contained in that section before the update (83), plus or minus the number of rows inserted or deleted from that section (0 inserted, 4 deleted) and plus or minus the number of rows moved into or out of that section (0 moved in, 0 moved out). with userInfo (null)
CoreData: error: Serious application error.  An exception was caught from the delegate of NSFetchedResultsController during a call to -controllerDidChangeContent:.  Invalid update: invalid number of rows in section 0.  The number of rows contained in an existing section after the update (83) must be equal to the number of rows contained in that section before the update (83), plus or minus the number of rows inserted or deleted from that section (0 inserted, 4 deleted) and plus or minus the number of rows moved into or out of that section (0 moved in, 0 moved out). with userInfo (null)
2016-12-18 18:23:49.612389 WordPress[2598:799241] Finished Notifications Background Fetch!
2016-12-18 18:23:49:612 WordPress[2598:792178] Finished Notifications Background Fetch!
```

**Comment by koke** [comment]
That makes sense. `Notification.sectionIdentifier` will change when the calendar day changes but won't generate any change notifications as the actual managed object doesn't change. Is it normal that the results controller is listening to updates when the app is launched in the background? I'd expect it to stop listening and reload the table data when the app becomes active

**Comment by koke** [comment]
I've spent a few hours trying to reproduce this but I haven't seen it once with the steps above :(

**Comment by jleandroperez** [comment]
@koke it's weird, since... i couldn't trigger it 10 over 10. There has to be yet another factor involved (potentially a race condition internally in the SDK?).

Anyways. This fellow never (FRC) stopped listening to Core Data changes, even when the app is in background. Sounds like a potential band aid patch (because... ideally we wouldn't need to do that right?).

But honestly, i can't think of another way to prevent that from happening.

**Comment by koke** [comment]
I found it surprising that the FRC kept listening for changes and updating the table while in the background. I'm not sure if that's expected behavior but it didn't feel right.

**Comment by jleandroperez** [comment]
Awesome work @koke!!!

**Comment by jleandroperez** [comment]
Reopening this one since it was reported again, in the latest internal beta.

**Comment by aerych** [comment]
A description of what happened as reported by Biser:

> I was able to reproduce this issue or something very similar to it on Saturday with the AppStore version of the app:
> 
> I attached the then current.log in 1a1fe-pb (with enabled extra debug in the settings). Relevant part starts at line 413, 2017/06/10 08:13:39:128.
> 
> Phone specs: iPhone 7+, 256G, iOS 10.3.3 ( I’m on the Public Beta system profile if that will help in any way )
> Note: I noticed that the WiFi connectivity in the room where we were was a bit spotty, so it’s possible it disconnected/connected at odd times. On the other hand mobile data is solid, so there shouldn’t have been a case of no internet connection, except from switching from/to WiFi.
>
> Steps that I followed that lead me to noticing it happening again (it hasn’t happened for a while now).
> * On Saturday morning I woke up and saw that there were several notifications on my lock screen, but the phone unlocked too quickly to see what they were
> * I saw there was a badge with ( 1 ) on the app
> * Opened the app
> * Saw that the new notifications weren’t being shown ( only one unread from the previous day )
> * Refreshed several times, but still nothing
> * Sent the app to background
> * Opened again
> * Refreshed several more times, nothing
> * I clicked the notification and I was surprised that if I opened the latest notification ( about post likes ) it showed the new events properly, but if I returned to the list of notifications it still showed the old state ( see attached screenshots )
> * Did several more refreshes – same result
> * Killed the app
> * Opened it again
> * Refreshed manually
> * The notifications properly “synced” and were showing the latest state

**Comment by rachelsquirrel** [comment]
I've got a report of this in 206275-HS and it has also happened to me several times. I have noticed that sometimes I'll receive an email notification on my iPhone letting me know I've got a new comment but there's nothing in the app. It's intermittent and I'm not sure what the pattern is for me but I'll comment again when it next happens.

**Comment by catehstn** [comment]
Another report that was likely this issue.

**Comment by charliescheer** [comment]
It looks like this problem is still happening.  We have been working on #1508240-zen where a user is describing a similar problem of the notifications panel being delayed.  Reopening.

**Comment by aerych** [comment]
I have seen something similar to this a couple of times. Each time I've meant to open an issue and each time I've forgotten before I could get back to my computer. Mea culpa. :/  Here's the behavior I've observed in case its relevant.

- Be on the notifications tab then background the app.
- Let enough time pass for the app to suspend and terminate. Use some other memory intensive application to encourage this.  Do not force terminate the app as we need to preserve state restoration for the glitch to present.
- Receive a few notifications.  Swipe to open the app via one of the notifications.
- When the app loads the notifications tab does not show the new notifications. 
- Background the app and then force close to reset state restoration. 
- Reopen the app.  Now the notifications list appears as expected.

Speculation: I have a hunch that there's something weird happening with state restoration when the app is launching in response to a notification.  Its almost like there are two copies of the notifications view controller instantiated and the one with the old state is overlaying the new one.  The reason I say this is due to some scrolling behavior I noticed during one or two of the incidents.  The notifications vc showed about one screen's worth of notification cells but I could scroll and scroll for several pages worth (no cells just empty space).  After force closing and returning I noted the notifications vc scrolled for about the same amount as before but this time with the correct content.  It made me wonder if the one vc's view was the content for the other vc's scroll view -- like it was a weird state restoration glitch.

**Comment by stevebaranski** [comment]
@aerych I've devoted some time to investigating this issue, and I wanted to share a few observations. I'd welcome your thoughts.

_State Restoration_

> I have a hunch that there's something weird happening with state restoration when the app is launching in response to a notification. 

This seems plausible to me. It is worth noting, however, that the original report (December 2016) preceded the implementation of state restoration in [February 2017](https://github.com/wordpress-mobile/WordPress-iOS/pull/6681).

Would it be feasible to disable state restoration for a time (e.g., a release or two) to evaluate its impact on the issue?

_Other Thoughts_

I've observed a bit of complexity related to preventing ["overwork"](https://github.com/wordpress-mobile/WordPress-iOS/blob/a177e20b65974c436364d68adb95eaf01578977c/WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.swift#L735-L764) that may be less necessary now that it was when this view controller was originally implemented.

Some of the checks with regard to `lastReloadDate` (which appears to influence FRC reloads) and `lastSeenTime` (which interacts with the `notifications/seen` endpoint) warrant some additional scrutiny.

Some of this logic comes into play when the app is [foregrounded](https://github.com/wordpress-mobile/WordPress-iOS/blob/a177e20b65974c436364d68adb95eaf01578977c/WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.swift#L472-L481). I've also noted that some of the aforementioned values are reset on [account change](https://github.com/wordpress-mobile/WordPress-iOS/blob/a177e20b65974c436364d68adb95eaf01578977c/WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.swift#L483-L489). This might be helpful if / when the issue occurs in the field.
